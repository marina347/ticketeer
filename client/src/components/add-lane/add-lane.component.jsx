import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import FormButton from "../form-button/form-button.component";
import FormInput from "../form-inputs/form-input/form-input.component";
import { AddLaneContainer } from "../add-lane/add-lane.styles";
import { addLaneAsync } from "../../redux/lane/lane.actions";
import { selectToken } from "../../redux/user/user.selectors";

export class AddLane extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      laneName: "",
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }
  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleSubmit(event) {
    const { boardId, addLane, token } = this.props;
    event.preventDefault();
    addLane(this.state.laneName, boardId, token);
    this.setState({ laneName: "" });
  }

  render() {
    return (
      <AddLaneContainer>
        <form id="add_lane_form" onSubmit={this.handleSubmit}>
          <FormInput
            id="add_lane_input"
            label="Add lane"
            type="text"
            name="laneName"
            placeholder="Lane name"
            value={this.state.laneName}
            handleChange={this.handleChange}
          />
          <FormButton type="submit">ADD</FormButton>
        </form>
      </AddLaneContainer>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  token: selectToken,
});

const mapDispatchToProps = (dispatch) => ({
  addLane: (name, boardId, token) =>
    dispatch(addLaneAsync(name, boardId, token)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddLane);
