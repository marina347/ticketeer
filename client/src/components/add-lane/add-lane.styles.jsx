import styled from "styled-components";

export const AddLaneContainer = styled.div`
  text-align: center;
  margin: 50px;
`;