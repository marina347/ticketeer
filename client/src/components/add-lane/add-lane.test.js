import React from "react";
import { shallow } from "enzyme";

import { AddLane } from "./add-lane.component";

describe("AddBoard component", () => {
  let wrapper;
  let mockProps;
  let addLane;
  let event;
  let handleSubmitSpy;

  beforeEach(() => {
    addLane = jest.fn();
    event = {
      preventDefault: jest.fn(),
      target: {
        name: "laneName",
        value: "testName",
      },
    };
    mockProps = {
      addLane,
    };

    //spyes will work if in class declaration you use bind in constructor. on anonymous method spyes cant work
    handleSubmitSpy = jest.spyOn(AddLane.prototype, "handleSubmit");

    wrapper = shallow(<AddLane {...mockProps} />);
  });

  it("Should render AddLane component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Should start onSubmit method when form inside is submitted", () => {
    wrapper
      .find("AddLaneContainer")
      .find('[id="add_lane_form"]')
      .simulate("submit", event);
    expect(handleSubmitSpy).toHaveBeenCalled();
    expect(addLane).toHaveBeenCalled();
    expect(wrapper.state()).toEqual({ laneName: "" });
  });

  it("Should set lanaName to testName if handleChange is fired", () => {
    wrapper.instance().handleChange(event);
    expect(wrapper.state()).toEqual({
      laneName: "testName",
    });
  });

  it("Form input has a prop handleChange", () => {
    expect(
      wrapper
        .find("AddLaneContainer")
        .find('[id="add_lane_form"]')
        .find('[id="add_lane_input"]')
        .prop("handleChange")
    ).not.toEqual(undefined);
  });
});
