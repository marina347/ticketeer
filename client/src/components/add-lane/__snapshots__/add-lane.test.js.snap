// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLane component Should render AddLane component 1`] = `
<form
  className="add-form"
  id="add_lane_form"
  onSubmit={[Function]}
>
  <h3
    className="add-form__heading"
  >
    ADD NEW LANE
  </h3>
  <FormInput
    handleChange={[Function]}
    id="add_lane_input"
    label="Lane name"
    maxlength="40"
    name="laneName"
    placeholder="Lane name"
    required={true}
    type="text"
    value=""
  />
  <FormButton
    className="btn btn-main btn-main--gradient u-animation-none u-margin-top-medium"
    type="submit"
  >
    ADD
  </FormButton>
</form>
`;
